osu.Game/Tests/Visual/ModPerfectTestScene.cs

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Game.Beatmaps;
+using osu.Game.Rulesets;
+using osu.Game.Rulesets.Mods;
+using osu.Game.Rulesets.Objects;
+
+namespace osu.Game.Tests.Visual
+{
+    public abstract class ModPerfectTestScene : ModTestScene
+    {
+        private readonly Ruleset ruleset;
+        private readonly ModPerfect mod;
+
+        protected ModPerfectTestScene(Ruleset ruleset, ModPerfect mod)
+            : base(ruleset)
+        {
+            this.ruleset = ruleset;
+            this.mod = mod;
+        }
+
+        protected void CreateHitObjectTest(HitObjectTestCase testCaseData, bool shouldMiss) => CreateModTest(new ModTestData
+        {
+            Mod = mod,
+            Beatmap = new Beatmap
+            {
+                BeatmapInfo = { Ruleset = ruleset.RulesetInfo },
+                HitObjects = { testCaseData.HitObject }
+            },
+            Autoplay = !shouldMiss,
+            PassCondition = () => ((PerfectModTestPlayer)Player).CheckFailed(shouldMiss && testCaseData.FailOnMiss)
+        });
+
+        protected override TestPlayer CreateModPlayer(Ruleset ruleset) => new PerfectModTestPlayer();
+
+        private class PerfectModTestPlayer : TestPlayer
+        {
+            public PerfectModTestPlayer()
+                : base(showResults: false)
+            {
+            }
+
+            protected override bool AllowFail => true;
+
+            public bool CheckFailed(bool failed)
+            {
+                if (!failed)
+                    return ScoreProcessor.HasCompleted && !HealthProcessor.HasFailed;
+
+                return HealthProcessor.HasFailed;
+            }
+        }
+
+        protected class HitObjectTestCase
```

Comment:
i still don't like this and think it's convoluted. should look at a better solution before merging (even if that means combining with the other "info" class in the parent test scene)

Reply:
just going to go with this for now (just renamed to match `ModTestScene`). i think we can refactor/rethink as this continues to get used in more places.